### PR TITLE
Redirect to login if trying to access another user private pages

### PR DESCRIPTION
**Pull request by ivanmalagon**
This PR redirects to login if trying to access another user private static pages.

### CODE REVIEW

File: lib/assets/javascripts/dashboard/helpers/redirector.js [hunk 1]
```diff
@@ -9,5 +9,12 @@ module.exports = {
       const newPathname = `/u/${username}/${page}`;
       location.replace(`${newOrigin}${newPathname}`);
     }
+  },
+
+  redirectToLoginIfAnotherIsLogged: function (username, location) {
+    const domains = parseDomains(location.href);
+    if (domains.subdomain && username && domains.subdomain !== username) {
```
Comment (javitonino) [inline lib/assets/javascripts/dashboard/helpers/redirector.js]: I think this condition is wrong. If I read this OK, this redirects if the subdomain != username, which is correct.

The algorithm needs to be something like this (hoping I didn't mess up).

1. Extract the subdomain and username (`/u/username` or `/user/username`) from the URL.
2. Extract session data (`userData.organization.name` and `userData.username`)
3. Extract if we are in subdomainless mode (we may need to add it to `/me`)
4. Subdomainless:
  - Ensure `url.username` = `session.username` or redirect to `/login`
5. Domainless, if `session.org` is present
  - Ensure `url.username` = `session.username` or redirect to `/login`
  - Ensure `url.subdomain` = `session.org` or fix the URL (redirect to `org.carto.com/u/username/xxxx`)

**Comment by ivanmalagon** [comment]
Pasting @javitonino comment that has disappeared because of a code change.

I think this condition is wrong. If I read this OK, this redirects if the subdomain != username, which is incorrect.

The algorithm needs to be something like this (hoping I didn't mess up).

1. Extract the subdomain and username (`/u/username` or `/user/username`) from the URL.
2. Extract session data (`userData.organization.name` and `userData.username`)
3. Extract if we are in subdomainless mode (we may need to add it to `/me`)
4. Subdomainless:
  - Ensure `url.username` = `session.username` or redirect to `/login`
5. Domainful, if `session.org` is present
  - Ensure `url.username` = `session.username` or redirect to `/login`
  - Ensure `url.subdomain` = `session.org` or fix the URL (redirect to `org.carto.com/u/username/xxxx`)
6. Domainful, if `session.org` is null
  - Ensure `url.subdomain` = `session.username` or redirect to `/login`
  - Ensure `url.username` is empty or fix the URL (redirect without the `/u/username` part)

I'd recommend to check this against an older, non-static, version of the code.

**Comment by ivanmalagon** [comment]
As you say, @javitonino we don't have currently any way of knowing if we are in subdomainless mode. Could you add that info to the endpoint? You can use this same branch.

Thank you for your thorough checklist.

**Comment by ivanmalagon** [comment]
What's thee source of `base_url`? Is it related with the request or is it something related to configuration?
In a domainless installation, what would be its value?

**Comment by javitonino** [comment]
It is generated from the session and configuration, taking into account subdomainless/domainful and org/non-org. it will always be the correct URL.

Examples:

subdomainless:
carto.com/user/username

domainful:
organization.carto.com/u/username
username.carto.com

**Comment by jesusbotella** [comment]
So, we can assume that if the `base_url` property that comes from `/api/v3/me` is different to the current URL except the path, the user needs to be redirected to login, right?

**Comment by jesusbotella** [comment]
I have just tested this PR in staging, it works successfully. There's one caveat though.

We will be able to redirect any user to the correct url as long as the subdomain is valid (i.e. the user/org in the subdomain exists), because NGINX validates it against redis. If it is not valid, it redirects the browser to `/noneuser`.

So, the cases are:

- If the user is not logged in, it will be redirected to `/login` at all times.

#### Organization URLs:
- If the username contained in `http://organization.carto.com/u/username` does not correspond to the logged user, the user will be redirected to `http://organization.carto.com/u/loggeduser/{page}`.
- If the organization contained in `http://organization.carto.com/u/username` does not correspond to the one of the logged user, the user will be redirected to `http://organizationthatuserbelongsto.carto.com/u/username/{page}`.
- If both are wrong, the user will be redirected to `http://organizationthatuserbelongsto.carto.com/u/loggeduser/{page}`

#### Regular URLs:
- If the username contained in `http://username.carto.com/` does not correspond to the logged user, the user will be redirected to `http://loggeduser.carto.com/{page}`.

I might had forgotten some cases, but every use case should work without errors ™️. It is currently deployed in staging cloud for you to test. If it doesn't work, the deploy might have been destroyed by somebody. 😂

File: lib/assets/javascripts/cartodb/helpers/redirector.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      return;
+    }
+    if (sessionOrg && urlSubdomain === sessionUsername) {
+      var newUrl = base_url + '/u/' + sessionUsername + '/' + page;
+      location.replace(newUrl);
+    }
+    location = '/login';
+  }
+
+// If the URL matches base_url -> OK
```
Comment (ivanmalagon) [inline lib/assets/javascripts/cartodb/helpers/redirector.js]: Delete these comments.

File: lib/assets/javascripts/dashboard/helpers/redirector.js [hunk 3]
```diff
@@ -1,13 +1,27 @@
...
-      const newOrigin = location.origin.replace(domains.subdomain, organization.name);
-      const newPathname = `/u/${username}/${page}`;
-      location.replace(`${newOrigin}${newPathname}`);
+  needsToChangeBaseURL: function (baseURL, location) {
+    var locationObject = location || window.location;
+    var currentURLOrigin = locationObject.origin;
+    var urlData = getUserPathname(locationObject.pathname);
+
+    if (urlData.username) {
+      return baseURL !== `${currentURLOrigin}${urlData.userPathname}`;
```
Comment (ivanmalagon) [inline lib/assets/javascripts/dashboard/helpers/redirector.js]: ES6!

**Pull request merged**